### PR TITLE
Recycling Recipes

### PR DESCRIPTION
**Pull request by DStrand1**

### CODE REVIEW

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 1]
```diff
@@ -1,111 +1,400 @@
...
+        return temp.entrySet().stream()
+                .sorted(Comparator.comparingLong(e -> -e.getKey().amount))
+                .limit(maxOutputs)
+                .map(Entry::getValue)
+                .collect(Collectors.toList());
+    }
+
+    private static void splitStacks(List<Tuple<ItemStack, MaterialStack>> list, ItemStack originalStack, UnificationEntry entry) {
+        int amount = originalStack.getCount();
+        while (amount > 64) {
```
Comment (DStrand1) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:338]: This *could* cause problems, but since we are pretty much only using a permutation of a Dust or Ingot, the stack size should be consistently 64

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 2]
```diff
@@ -1,111 +1,400 @@
...
+            return new MaterialStack(Materials.Ash, amount / 8);
+        }
+
+        // Else if the Material is a Gem, process its output (see below)
+        if (material.hasProperty(PropertyKey.GEM)) {
+            return getGemArcSmeltResult(materialStack);
+        }
+
+        // Else if the Material has NO_SMELTING, return nothing
+        if (material.hasFlag(NO_SMELTING)) {
```
Comment (DStrand1) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:181]: This prevents things like Brick Dust and Stone Dust from being output in the process

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 3]
```diff
@@ -1,111 +1,400 @@
...
 import static gregtech.api.GTValues.M;
-import static gregtech.api.unification.material.info.MaterialFlags.FLAMMABLE;
+import static gregtech.api.unification.material.info.MaterialFlags.*;
 
 public class RecyclingRecipes {
 
-    public static void init() {
-        initializeArcRecyclingRecipes();
-    }
+    // TODO - Fix recipe order with some things (noticed Hermetic Casings)
```
Comment (DStrand1) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:31]: This should be addressed, but there might be an edge case still. Loading order is important for some recipes now, as if a later recipe uses an input (say, Steam Oven using Bronze Casings), unification data could be wrong

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 4]
```diff
@@ -1,111 +1,400 @@
...
+    // TODO - Figure out solution to LuV+ components
```
Comment (DStrand1) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:32]: Discussed, and will be in a later PR

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 5]
```diff
@@ -1,111 +1,400 @@
...
+    // TODO - (to remember) Do NOT calculate any material component lists for circuits, they are simply totally lost
+    // TODO - Work on durations and EUt's
```
Comment (DStrand1) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:34]: The durations are resolved, though EU/t's may be too high for some recipes

File: src/main/java/gregtech/loaders/MaterialInfoLoader.java [hunk 6]
```diff
@@ -82,6 +83,56 @@ public static void init() {
...
+        OreDictUnifier.registerOre(MetaTileEntities.HULL[8].getStackForm(), new ItemMaterialInfo(
+                new MaterialStack(Materials.Livermorium, OrePrefix.plate.materialAmount * 8),
+                new MaterialStack(Materials.NaquadahAlloy, OrePrefix.cableGtSingle.materialAmount * 2),
+                new MaterialStack(Materials.Rubber, OrePrefix.plate.materialAmount * 2)));
+
+        OreDictUnifier.registerOre(MetaTileEntities.HULL[14].getStackForm(), new ItemMaterialInfo(
+                new MaterialStack(Materials.Neutronium, OrePrefix.plate.materialAmount * 8),
+                new MaterialStack(Materials.RutheniumTriniumAmericiumNeutronate, OrePrefix.cableGtSingle.materialAmount * 2),
+                new MaterialStack(Materials.Rubber, OrePrefix.plate.materialAmount * 2)));
+
```
Comment (DStrand1) [inline src/main/java/gregtech/loaders/MaterialInfoLoader.java:135]: These had to be done separately since the handcrafting recipe is more resource-expensive than the Assembler recipe, leading to potential dupes. If we add a method onto `RecipeBuilder` to calculate unification data, these can be removed (as can the entries for EBF coils)

File: src/main/java/gregtech/api/unification/OreDictUnifier.java [hunk 7]
```diff
@@ -266,21 +274,40 @@ else if ((materialAmount * 9) >= M)
...
 
+    public static ItemStack getIngot(MaterialStack materialStack) {
+        return getIngot(materialStack.material, materialStack.amount);
+    }
+
+    /**
+     * Returns an Ingot of the material if it exists. Otherwise it returns a Dust.
+     * Returns ItemStack.EMPTY if neither exist.
+     */
+    public static ItemStack getIngotOrDust(Material material, long materialAmount) {
```
Comment (DStrand1) [inline src/main/java/gregtech/api/unification/OreDictUnifier.java:301]: Needed to not return an `ItemStack.EMPTY` for some Arc Furnace recipes (Ash, in particular)

File: src/main/java/gregtech/api/unification/ore/OrePrefix.java [hunk 8]
```diff
@@ -368,6 +368,12 @@
         pipeNormalRestrictive.addSecondaryMaterial(new MaterialStack(Materials.Iron, ring.materialAmount * 2));
         pipeLargeRestrictive.addSecondaryMaterial(new MaterialStack(Materials.Iron, ring.materialAmount * 2));
 
+        cableGtSingle.addSecondaryMaterial(new MaterialStack(Materials.Rubber, plate.materialAmount));
+        cableGtDouble.addSecondaryMaterial(new MaterialStack(Materials.Rubber, plate.materialAmount));
+        cableGtQuadruple.addSecondaryMaterial(new MaterialStack(Materials.Rubber, plate.materialAmount * 2));
+        cableGtOctal.addSecondaryMaterial(new MaterialStack(Materials.Rubber, plate.materialAmount * 3));
+        cableGtHex.addSecondaryMaterial(new MaterialStack(Materials.Rubber, plate.materialAmount * 4));
```
Comment (DStrand1) [inline src/main/java/gregtech/api/unification/ore/OrePrefix.java:375]: These are not quite accurate to their current costs, but a PR reworking cable covering will come shortly after this (before release)

File: src/main/java/gregtech/api/recipes/ModHandler.java [hunk 9]
```diff
@@ -337,6 +345,83 @@ public static Object finalizeIngredient(Object ingredient) {
...
+        }
+
+        char lastChar = ' ';
+        for (int i = itr; i < recipe.length; i++) {
+            Object ingredient = recipe[i];
+
+            // Track the current working ingredient symbol
+            if (ingredient instanceof Character) {
+                lastChar = (char) ingredient;
+                continue;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/ModHandler.java:370]: Why does this continue after finding the current working symbol? Shouldn't it allow following down the loop to get to the other parts instead of `continue`?

File: src/main/java/gregtech/api/recipes/ModHandler.java [hunk 10]
```diff
@@ -337,6 +345,83 @@ public static Object finalizeIngredient(Object ingredient) {
...
+            }
+
+            // Should never happen if recipe is formatted correctly
+            // In the case that it isn't, this error should be handled
+            // by an earlier method call parsing the recipe.
+            if (lastChar == ' ') return null;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/ModHandler.java:376]: What about recipe objects that have a `' '` in them? Something like 

```
'WC'
'S '
```
To create a 2x2 recipe with the bottom right space being empty. Or something in the shape of the armor chestplate, which has the top middle slot as a `' '`.

File: src/main/java/gregtech/api/recipes/ModHandler.java [hunk 11]
```diff
@@ -337,6 +345,83 @@ public static Object finalizeIngredient(Object ingredient) {
...
+
+            // First try to get ItemMaterialInfo
+            ItemMaterialInfo info = OreDictUnifier.getMaterialInfo(stack);
+            if (info != null) {
+                for (MaterialStack ms : info.getMaterials()) func.accept(ms, lastChar);
+                continue;
+            }
+
+            // Then try to get a single Material (UnificationEntry needs this, for example)
+            MaterialStack materialStack = OreDictUnifier.getMaterial(stack);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/ModHandler.java:406]: Why not `info.getMaterial()`? Since `OreDictUnifier.getMaterial(stack)` basically ends up calling this anyway.

File: src/main/java/gregtech/api/unification/ore/OrePrefix.java (hunk 8 above)
Comment (ALongStringOfNumbers) [reply src/main/java/gregtech/api/unification/ore/OrePrefix.java:375]: Now that the PR to rework cable covering is in, are these values accurate to the current costs?

Comment (DStrand1) [reply src/main/java/gregtech/api/unification/ore/OrePrefix.java:375]: The `cableGtHex` value is incorrect, which I will fix (should be 5)

File: src/main/java/gregtech/api/recipes/ModHandler.java (hunk 10 above)
Comment (DStrand1) [reply src/main/java/gregtech/api/recipes/ModHandler.java:376]: This is in the section parsing the `key, value` pairs for the recipe nputs, not the pattern. Empty spaces in the pattern will pass through fine

File: src/main/java/gregtech/api/recipes/ModHandler.java (hunk 11 above)
Comment (DStrand1) [reply src/main/java/gregtech/api/recipes/ModHandler.java:406]: For whatever reason, some unification data is stored in `materialUnificationInfo`, and others are stored in `stackUnificationInfo`. These two methods of attempting to acquire unification data simply query both options.

We cannot simply call `getMaterial()`, since if it had an `ItemMaterialInfo` with multiple entries, we would only get the first Material back instead of all of them

File: src/main/java/gregtech/api/recipes/ModHandler.java (hunk 9 above)
Comment (DStrand1) [reply src/main/java/gregtech/api/recipes/ModHandler.java:370]: It's just how it ended up laying out, alternatively I could iterate over `i += 2` in the loop, but either work fine

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 12]
```diff
@@ -1,111 +1,400 @@
...
+        // Calculate the duration based off of those two possible outputs.
+        // - Sum the two Material amounts together (if both exist)
+        // - Divide the sum by M
+        long duration = fluidMs.amount * fluidMs.material.getAverageMass();
+        if (itemMs != null) duration += (itemMs.amount * itemMs.material.getAverageMass());
+        duration = Math.max(1L, duration / M);
+
+        // Build the final Recipe.
+        RecipeBuilder<?> extractorBuilder = RecipeMaps.EXTRACTOR_RECIPES.recipeBuilder()
+                .inputs(input)
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:118]: `input.copy()` like the macerator?

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 13]
```diff
@@ -1,111 +1,400 @@
...
+                RecipeMaps.ARC_FURNACE_RECIPES.getMaxOutputs(),
+                RecyclingRecipes::getArcIngotOrDust
+        );
+
+        // Exit if no valid outputs exist for this recycling Recipe.
+        if (outputs.size() == 0) return;
+
+        // Build the final Recipe.
+        RecipeMaps.ARC_FURNACE_RECIPES.recipeBuilder()
+                .inputs(input)
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:155]: `input.copy()` like the macerator?

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 14]
```diff
@@ -1,111 +1,400 @@
...
     }
 
-    private static ItemStack getArcSmeltingResult(MaterialStack materialStack) {
+    private static MaterialStack getArcSmeltingResult(MaterialStack materialStack) {
         Material material = materialStack.material;
-        long materialAmount = materialStack.amount;
+        long amount = materialStack.amount;
+
+        if (material.hasFlag(EXPLOSIVE)) {
+            return new MaterialStack(Materials.Ash, amount / 16);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:167]: Should these return at minimum 1, to ensure that all the inputs would be accounted for, or are you find returning a materialStack with size 0 if `amount` is less than your divisor?

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 15]
```diff
@@ -1,111 +1,400 @@
...
+    }
+
+    private static MaterialStack getGemArcSmeltResult(MaterialStack materialStack) {
+        Material material = materialStack.material;
+        long amount = materialStack.amount;
+
+        // If the Gem Material has Oxygen in it, return Ash
+        if (material.getMaterialComponents().stream()
+                .anyMatch(stack -> stack.material == Materials.Oxygen)) {
+            return new MaterialStack(Materials.Ash, amount / 8);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:210]: Same here with the minimum 1 question

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 16]
```diff
@@ -1,111 +1,400 @@
...
+        // This can sometimes be before the Fluid output in the list, so we have to
+        // assume it can be anywhere in the list.
+        MaterialStack itemMs = materials.stream().filter(ms -> !ms.material.equals(fluidMs.material)).findFirst().orElse(null);
+
+        // Calculate the duration based off of those two possible outputs.
+        // - Sum the two Material amounts together (if both exist)
+        // - Divide the sum by M
+        long duration = fluidMs.amount * fluidMs.material.getAverageMass();
+        if (itemMs != null) duration += (itemMs.amount * itemMs.material.getAverageMass());
+        duration = Math.max(1L, duration / M);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:114]: Can't this just be passed into `calculateDuration` if you add some simple null checking to that method? eg: `if(material == null) continue;`

**Review by ALongStringOfNumbers — COMMENTED** [review]
Magnetic Materials Arc smelt into their non-magnetic form, but macerate into the magnetic dust. Is this intended?

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java (hunk 14 above)
Comment (DStrand1) [reply src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:167]: The material stack amount after dividing should never reasonably be zero here, at worst it would be something like `OrePrefix.dustTiny.materialAmount / 16`, or `403200 / 16` = 25,200 (this would still not create any output, but that is fine in a lot of cases)

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java (hunk 16 above)
Comment (DStrand1) [reply src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:114]: I could, but honestly I think how it is here is as clean as it can get in the code

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java [hunk 17]
```diff
@@ -1,111 +1,402 @@
...
+
+        // Exit if no valid Materials exist for this recycling Recipe.
+        if (outputs.size() == 0) return;
+
+        // Build the final Recipe.
+        RecipeMaps.MACERATOR_RECIPES.recipeBuilder()
+                .inputs(input.copy())
+                .outputs(outputs)
+                .duration(calculateDuration(materials))
+                .EUt(4 * multiplier)
```
Comment (kumquat-ir) [inline src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:95]: Macerator default EU/t has been changed to 2 instead of 4, as that is what it is in 5u
(also, extractor and arc EU/t should probably be changed to VA[LV])

File: src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java (hunk 3 above)
Comment (ALongStringOfNumbers) [reply src/main/java/gregtech/loaders/recipe/RecyclingRecipes.java:32]: I have not noticed issues with the Hermetic Casings when I check

File: src/main/java/gregtech/loaders/MaterialInfoLoader.java (hunk 6 above)
Comment (ALongStringOfNumbers) [reply src/main/java/gregtech/loaders/MaterialInfoLoader.java:119]: Minor nitpick that should not really matter. These should use the proper assembler fluid materials, instead of all just using rubber

File: src/main/java/gregtech/api/unification/OreDictUnifier.java [hunk 18]
```diff
@@ -267,21 +275,40 @@ else if ((materialAmount * 9) >= M)
...
+        ItemStack ingotStack = getIngot(material, materialAmount);
+        if (ingotStack != ItemStack.EMPTY) return ingotStack;
+        return getDust(material, materialAmount);
+    }
+
+    public static ItemStack getIngotOrDust(MaterialStack materialStack) {
+        return getIngotOrDust(materialStack.material, materialStack.amount);
+    }
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/unification/OreDictUnifier.java:310]: Why did you add this method, when anything that can call this method can automatically call the method's redirect, as it has access to the `materialStack.amount` and `materialStack.material` already. This seems like a useless method

**Review by ALongStringOfNumbers — APPROVED** [review]
Other than my minor comments, this looks good to me
